lex.go

```diff
@@ -370,32 +370,33 @@ func lexExpr(l *lexer) stateFn {
 
 // lexName is the state function for lexing names.
 func lexName(l *lexer) stateFn {
-	var s string
+	var typ itemType
 
-	typ := itemUnquotedLiteral
+	switch l.next() {
+	case '$':
+		typ = itemVariable
+	case '.':
+		typ = itemKeyword
+	default:
+		typ = itemUnquotedLiteral
 
-	for {
-		r := l.next()
+		l.backup() // backup to the first rune
+	}
 
-		switch {
-		default:
-			l.backup()
+	var (
+		s string // item value
+		r rune   // current rune
+	)
 
-			return l.emitItem(mk(typ, s))
-		case len(s) > 0 && isName(r):
-			s += string(r)
-		case r == eof:
-			return l.emitItem(mk(typ, s))
-		case len(s) == 0 && isNameStart(r):
-			s = string(r)
-		case len(s) == 0 && r == '$':
-			s = string(r)
-			typ = itemVariable
-		case len(s) == 0 && r == '.':
-			s = string(r)
-			typ = itemKeyword
-		}
+	for r = l.next(); isName(r); r = l.next() {
+		s += string(r)
 	}
+
+	if r != eof { // backup is not necessary if we reached eof
```

Comment:
Isn't `eof` broken MF2 syntax?

Name is either a variable name or an unquoted literal.

Reply:
it is, but it has same behaviour as it was before

Reply:
Let's do it. Let's return the unexpected eof error.